### PR TITLE
New datasets, cyclic cosine and bug fixes

### PR DESCRIPTION
**Pull request by caiom**
1. Add new datasets: 'olx_WordData20210110', 'olx_OutlookData20210917x2', 'olx_WordData20211003'
2. Adds cyclic cosine annealing
3. Fix bugs for the use of non-adaptive emb/softmax
4. Primer EZ separated into primer_conv and primer_sqrt
5. Fix onnx export config.json path
6. Removed DataFileStats for the train file as it takes too long and it is not used currently

### CODE REVIEW

File: archai/nlp/nvidia_transformer_xl/onnx/onnx_utils/load.py [hunk 1]
```diff
@@ -76,6 +76,11 @@ def load_from_pt(torch_model_path: str) -> Tuple[MemTransformerLM, dict]:
     checkpoint['model_config']['use_cache'] = True
 
     # Initializes the model
+
+    # Compatibility, TODO: Remove when not necessary
```
Comment (gugarosa) [inline archai/nlp/nvidia_transformer_xl/onnx/onnx_utils/load.py:80]: Should this still be necessary since we are re-training every model from scratch? They should not have this key anymore, right?

File: archai/nlp/nvidia_transformer_xl/nvidia_utils/exp_utils.py [hunk 2]
```diff
@@ -238,8 +238,8 @@ def dataset_dir_name(dataset:str)->str:
         return 'wikitext-103'
     if dataset=='wt2':
         return 'wikitext-2'
-    if dataset=='olx':
-        return 'olx'
+    if 'olx' in dataset:
```
Comment (gugarosa) [inline archai/nlp/nvidia_transformer_xl/nvidia_utils/exp_utils.py:241]: I would say that it might be interesting to change this function to a dictionary in the near future as the amount of datasets might probably increase. Accessing dictionaries should be less expensive than checking for "ifs".

File: archai/nlp/nvidia_transformer_xl/mem_transformer.py [hunk 3]
```diff
@@ -474,14 +474,14 @@ def forward(self, dec_inp, dec_attn_mask=None, mems=None, past_key_values=None):
...
+                 primer_conv=False, primer_sqrt=False, use_cache=False, **kwargs):
         super(RelLearnableDecoderLayer, self).__init__()
 
         self.dec_attn = RelLearnableMultiHeadAttn(n_head, d_model, d_head,
-                                                  dropout, primer_ez=primer_ez, use_cache=use_cache,
+                                                  dropout, primer_ez=primer_conv, use_cache=use_cache,
                                                   **kwargs)
 
-        if primer_ez:
+        if primer_sqrt:
```
Comment (gugarosa) [inline archai/nlp/nvidia_transformer_xl/mem_transformer.py:484]: Any particular reason for `sqrt` instead of `sqrd`/`squared`?

File: archai/nlp/nvidia_transformer_xl/nvidia_utils/corpus.py [hunk 4]
```diff
@@ -115,9 +115,16 @@ def _get_file_stats(filepath:str)->DataFileStats:
...
+    # def file_stats(self)->Tuple[DataFileStats, DataFileStats, DataFileStats]:
+    #     train_filepath, valid_filepath, test_filepath = self._dataset_filepaths()
+    #     return (Corpus._get_file_stats(train_filepath), \
+    #             Corpus._get_file_stats(valid_filepath), \
+    #             Corpus._get_file_stats(test_filepath))
+
     def file_stats(self)->Tuple[DataFileStats, DataFileStats, DataFileStats]:
         train_filepath, valid_filepath, test_filepath = self._dataset_filepaths()
-        return (Corpus._get_file_stats(train_filepath), \
+        return (None, \
```
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/nvidia_utils/corpus.py:127]: Why remove stats for train file?

File: archai/nlp/nvidia_transformer_xl/nvidia_utils/corpus.py [hunk 5]
```diff
@@ -138,7 +145,7 @@ def _create_vocab(datadir:str, dataset:str, vocab_type:str, vocab_cache_dir:str,
         if vocab_type == 'word':
             # '<S>' is added for double eos and <unk> is rare token in corpus with freq < 3
             bos_token, eos_token, lower_case, vocab_file = None, '<eos>', False, None # vocab file is text file of symbols, one per line
-            if dataset in ['wt103', 'wt2', 'olx']:
+            if dataset in ['wt103', 'wt2'] or 'olx' in dataset:
```
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/nvidia_utils/corpus.py:148]: I would suggest we use more restrictive filter such as starts_with('olx_')

File: archai/nlp/nvidia_transformer_xl/nvidia_utils/cyclic_cosine_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
+from collections.abc import Iterable
+from math import log, cos, pi, floor
+
+from torch.optim.lr_scheduler import _LRScheduler
+
+
```
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/nvidia_utils/cyclic_cosine_scheduler.py:6]: Could you please add comments on why we need this, where this code comes form and may be line or two of general overview?

File: archai/nlp/nvidia_transformer_xl/nvidia_utils/exp_utils.py (hunk 2 above)
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/nvidia_utils/exp_utils.py:241]: Same comment as previous for more restrictive check

File: archai/nlp/nvidia_transformer_xl/nvidia_utils/proj_adaptive_softmax.py [hunk 7]
```diff
@@ -52,9 +52,12 @@ def __init__(self, n_token, d_embed, d_proj,
...
 
-        self.shortlist_size = self.cutoffs[0]
-        self.n_clusters = len(self.cutoffs) - 1 # number of clusters will be >= 0
-        self.head_size = self.shortlist_size + self.n_clusters
+        if self.cutoffs:
+            self.n_clusters = len(self.cutoffs) - 1 # number of clusters will be >= 0
+            self.shortlist_size = self.cutoffs[0] # TODO: remove? Never used...
+            self.head_size = self.shortlist_size + self.n_clusters # TODO: remove? Never used...
+        else:
+            self.n_clusters = 0
```
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/nvidia_utils/proj_adaptive_softmax.py:60]: Wondering why we need this. I think the code handled the empty cutoffs already. No?

File: archai/nlp/nvidia_transformer_xl/nvidia_utils/proj_adaptive_softmax.py [hunk 8]
```diff
@@ -151,6 +154,11 @@ def default_tie_proj(cutoffs:List[int])->List[bool]:
     def clean_cutoffs(cutoffs:Optional[List[int]], n_token:int):
         if cutoffs is None:
             cutoffs = ProjectedAdaptiveLogSoftmax.default_cutoffs(n_token)
+
+        # Not using adaptive emb
+        if len(cutoffs) == 0:
```
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/nvidia_utils/proj_adaptive_softmax.py:159]: Why do we need this? It looks to me that empty cutoffs are handled properly. Instead of premature return, we return proper cutoff if cutoff was empty array.

File: archai/nlp/nvidia_transformer_xl/nvidia_utils/proj_adaptive_softmax.py [hunk 9]
```diff
@@ -183,7 +191,7 @@ def clean_tie_projs(tie_projs:Optional[List[bool]], cutoffs:List[int], n_token:i
         return tie_projs[:len(cutoffs)]
 
     def get_out_proj(self, i):
-        if self.tie_projs[i]:
+        if not self.tie_projs or self.tie_projs[i]:
```
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/nvidia_utils/proj_adaptive_softmax.py:194]: I thought this was handled. If tie_projs is empty it would be reset to proper value.

File: archai/nlp/nvidia_transformer_xl/olx_small.yaml [hunk 10]
```diff
@@ -0,0 +1,104 @@
+# Base
```
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/olx_small.yaml:1]: Wondering why we need new yaml. Can we inherit from existing sections and use existing yaml?

File: archai/nlp/nvidia_transformer_xl/onnx/onnx_utils/load.py [hunk 11]
```diff
@@ -76,6 +76,11 @@ def load_from_pt(torch_model_path: str) -> Tuple[MemTransformerLM, dict]:
...
+    if 'encoder_like' in checkpoint['model_config']:
```
Comment (sytelus) [inline archai/nlp/nvidia_transformer_xl/onnx/onnx_utils/load.py:81]: Could you please add comments what is this about?

**Review by sytelus — CHANGES_REQUESTED** [review]
